lib/mpl_toolkits/mplot3d/art3d.py

```diff
@@ -474,34 +462,43 @@ def set_3d_properties(self, zs, zdir):
             xs = []
             ys = []
         self._offsets3d = juggle_axes(xs, ys, np.atleast_1d(zs), zdir)
-        self._facecolor3d = self.get_facecolor()
-        self._edgecolor3d = self.get_edgecolor()
+        self._vzs = None
         self.stale = True
 
     @_api.delete_parameter('3.4', 'renderer')
     def do_3d_projection(self, renderer=None):
-        # see _update_scalarmappable docstring for why this must be here
-        _update_scalarmappable(self)
         xs, ys, zs = self._offsets3d
         vxs, vys, vzs, vis = proj3d.proj_transform_clip(xs, ys, zs,
                                                         self.axes.M)
-
-        fcs = (_zalpha(self._facecolor3d, vzs) if self._depthshade else
-               self._facecolor3d)
-        fcs = mcolors.to_rgba_array(fcs, self._alpha)
-        super().set_facecolor(fcs)
-
-        ecs = (_zalpha(self._edgecolor3d, vzs) if self._depthshade else
-               self._edgecolor3d)
-        ecs = mcolors.to_rgba_array(ecs, self._alpha)
-        super().set_edgecolor(ecs)
+        self._vzs = vzs
         super().set_offsets(np.column_stack([vxs, vys]))
 
         if vzs.size > 0:
             return min(vzs)
         else:
             return np.nan
 
+    def get_facecolor(self):
+        bc = super().get_facecolor()
+        fcs = (
+            _zalpha(bc, self._vzs)
+            if self._vzs is not None and self._depthshade
+            else bc
+        )
+        return mcolors.to_rgba_array(fcs, self._alpha)
+
+    def get_edgecolor(self):
+        if cbook._str_equal(self._edgecolors, 'face'):
+            return self.get_facecolor()
```

Comment:
Maybe add a comment why this is done here, instead of leaving it to the `super().get_edgecolor()`?